modules/core/src/main/scala/doobie/util/put.scala

```diff
@@ -56,25 +56,37 @@ sealed abstract class Put[A](
     ) {}
 
   def unsafeSetNonNullable(ps: PreparedStatement, n: Int, a: A): Unit =
-    if (a == null) sys.error("oops, null")
-    else put.fi.apply(ps, n, (put.k(a)))
+    unsafeSetWhatShouldNotBeNull(ps, n, a, expectedNonNullableParamAt(n))
 
   def unsafeSetNullable(ps: PreparedStatement, n: Int, oa: Option[A]): Unit =
     oa match {
-      case Some(a) => unsafeSetNonNullable(ps, n, a)
+      case Some(a) => unsafeSetWhatShouldNotBeNull(ps, n, a, expectedOptionalParamAt(n))
       case None    => unsafeSetNull(ps, n)
     }
 
+  private def unsafeSetWhatShouldNotBeNull(ps: PreparedStatement, n: Int, a: A, message: String): Unit =
+    if (a == null) sys.error(message)
+    else put.fi.apply(ps, n, (put.k(a)))
```

Comment:
In this case we're creating a new string for every value we set, which is probably not ideal for performance.
We probably want to only construct the message in the `if (a == null)` branch. To simplify things, I think it's probably not necessary to differentiate between `null` and `Option(null)` (though I appreciate the thoughtfulness behind it :slightly_smiling_face: )

Reply:
@jatcwang I have updated the code. It was the `n` param info that caused creating new string for every value we set. Now both strings are persisted and we keep the `null` vs `Option(null)` distinction :)

Reply:
Or do you think super simple would be even better? Like this --> https://github.com/typelevel/doobie/pull/2142

Reply:
Or with message constructed only in the `if (a == null)` branch and full information --> #2143